### PR TITLE
[determinism] Add GPU excepts, CPU d9m, and tests to crop_and_resize

### PR DESCRIPTION
**Pull request by duncanriach**
This current PR adds the following functionality, plus associated tests:

When `TF_DETERMINISTIC_OPS` is set to `"true"` or `"1"` (when op-determinism is expected),

1. the gradient w.r.t `image` of `tf.image.crop_and_resize` when running on CPU will be bit-exactly reproducible (from run-to-run), and
2. an attempt to use the GPU kernels for gradient w.r.t either `image` or `boxes` of `tf.image.crop_and_resize` will cause a `tf.errors.UnimplementedError` to be thrown along with an understandable message.

This current PR is associated with [RFC: Enabling Determinism in TensorFlow](https://github.com/tensorflow/community/blob/master/rfcs/20210119-determinism.md).

cc @sanjoy, @reedwm, @nluehr

### CODE REVIEW

File: tensorflow/python/ops/image_grad_test_base.py [hunk 1]
```diff
@@ -535,17 +536,28 @@ def crop_resize(image_tensor, boxes_tensor):
               with test_util.device(use_gpu=True):
                 with self.cached_session():
                   # pylint: disable=cell-var-from-loop
-                  err1 = gradient_checker_v2.max_error(
-                      *gradient_checker_v2.compute_gradient(
+                  if (self.__class__.__name__ ==
+                      "CropAndResizeOpDeterministicTest" and
```
Comment (reedwm) [inline tensorflow/python/ops/image_grad_test_base.py:540]: Instead of checking for the class name, I would check if `os.environ['TF_DETERMINISTIC_OPS']` is "1".

Also, maybe just disable this test when determinism is used, since you already test the backprop error messages elsewhere.

File: tensorflow/python/ops/image_grad_deterministic_test.py [hunk 2]
```diff
@@ -127,6 +135,164 @@ def resize_bilinear_gradients(local_seed):
...
+    boxes_shape = (num_boxes, 4)
+    random_seed.set_seed(123)
+    image = random_ops.random_normal(shape=image_shape, dtype=dtype)
+    boxes = random_ops.random_uniform(shape=boxes_shape, dtype=dtypes.float32)
+    box_indices = random_ops.random_uniform(
+        shape=(num_boxes,), minval=0, maxval=batch_size, dtype=dtypes.int32)
+    crop_size = constant_op.constant([3, 3], dtype=dtypes.int32)
+    return image, boxes, box_indices, crop_size
+
+  def _imageErrorMessage(self):
```
Comment (reedwm) [inline tensorflow/python/ops/image_grad_deterministic_test.py:163]: Make this a class constant:

```python
 _IMAGE_ERROR_MESSAGE = ("Deterministic GPU implementation of" +
                         " CropAndResizeBackpropImage not available")
```

And same for `_boxesErrorMessage`.

File: tensorflow/python/ops/image_grad_deterministic_test.py [hunk 3]
```diff
@@ -127,6 +135,164 @@ def resize_bilinear_gradients(local_seed):
...
+class CropAndResizeOpDeterminismExceptionsTest(test.TestCase):
+  """Test d9m-unimplemented exceptions from CropAndResizeBackprop{Image|Boxes}.
+
+  Test that tf.errors.UnimplementedError is thrown or not thrown, as
+  appropriate, by the GPU code-paths for CropAndResizeBackprop{Image|Boxes} when
+  deterministic ops are enabled.
+
+  This test assumes that the base op test runs all the same test cases when
+  deterministic ops are not enabled and will therefore detect erroneous
+  exception throwing in those cases.
```
Comment (reedwm) [inline tensorflow/python/ops/image_grad_deterministic_test.py:147]: I don't quite understand this comment. What is the "base op test"?

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/image_grad_deterministic_test.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+# Copyright 2020, 2021 The TensorFlow Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (duncanriach) [reply tensorflow/python/ops/image_grad_deterministic_test.py:1]: The PR does not change the year that the file was created.

See

> "When making changes to code with an existing copyright notice: ... Optionally, add current copyright year."

File: tensorflow/python/ops/image_grad_test_base.py (hunk 1 above)
Comment (duncanriach) [reply tensorflow/python/ops/image_grad_test_base.py:540]: I chose to do it this way to document the provenance of this conditional branch (i.e. in a child class) and to encapsulate the mechanism by which op-determinism is enabled. However, I don't see any major downsides to changing this, specifically:

  1. if this condition is removed by someone who doesn't understand why it's there or why it's needed then tests will fail and the person who made the change will be alerted to the error, and
  2. when `TF_DETERMINISTIC_OPS` is replaced with `tf.config.*deterministic*`, and `TF_DETERMINISTIC_OPS` is no longer set to '1' in the child class, then this test will fail because the exceptions will no longer be thrown.

> Also, maybe just disable this test when determinism is used, since you already test the backprop error messages elsewhere.

Testing that the functionality in this conditional branch is correct will ensure that when the exceptions are removed and deterministic GPU backprop functionality is added, this test will be modified to test that functionality. If the gradient is just skipped when determinism is expected then there is a higher probability that the deterministic GPU backprop functionality will (accidentally) not be tested.

This will be changed to `os.getenv('TF_DETERMINISTIC_OPS', '0') == '1'` in the next commit.

File: tensorflow/python/ops/image_grad_deterministic_test.py (hunk 2 above)
Comment (duncanriach) [reply tensorflow/python/ops/image_grad_deterministic_test.py:163]: Thanks for the simplification suggestion.

I originally put these messages in the class-level with the intention of using them in the parent class, but I ended up deciding to not test the error messages in the parent class; there it seemed sufficient to simply check that the correct exception type was being thrown. In the upcoming commit, I have simplified this so that there is nothing at all at the class level related to the exception messages.

File: tensorflow/python/ops/image_grad_deterministic_test.py (hunk 3 above)
Comment (duncanriach) [reply tensorflow/python/ops/image_grad_deterministic_test.py:147]: The upcoming commit, changes "base op test" to "test_base.CropAndResizeOpTestBase" in this particular instance. It also makes various similar clarifications elsewhere in `image_grad_deterministic_test.py`, `xent_op_deterministic_test.py`, and `sparse_xent_op_deterministic_test.py`.

**Review by reedwm — APPROVED** [review]
